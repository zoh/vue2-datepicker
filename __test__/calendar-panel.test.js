import { mount, shallowMount } from '@vue/test-utils';
import CalendarPanel from '../src/calendar/calendar-panel';

let wrapper;

afterEach(() => {
  wrapper.destroy();
});

describe('CalendarPanel', () => {
  it('feat: type = date, should emit select when click date', () => {
    wrapper = mount(CalendarPanel, {
      propsData: {
        value: new Date(2019, 9, 4),
      },
    });
    const tds = wrapper.findAll('.mx-table-date td');
    tds.at(0).trigger('click');
    expect(wrapper.emitted().select[0][0]).toEqual(new Date(2019, 8, 29));
  });

  it('feat: type = month, should emit event when click month', () => {
    wrapper = mount(CalendarPanel, {
      propsData: {
        type: 'month',
        defaultValue: new Date(2019, 9, 1),
      },
    });
    const tds = wrapper.findAll('.mx-table-month td > div');
    tds.at(0).trigger('click');
    expect(wrapper.emitted().select[0][0]).toEqual(new Date(2019, 0, 1));
  });

  it('feat: type = year, should emit event when click year', () => {
    wrapper = mount(CalendarPanel, {
      propsData: {
        type: 'year',
        defaultValue: new Date(2019, 9, 1),
      },
    });
    const tds = wrapper.findAll('.mx-table-year td > div');
    tds.at(0).trigger('click');
    expect(wrapper.emitted().select[0][0]).toEqual(new Date(2010, 0));
  });

  it('feat: when year >= 0 && year < 100, should be emit right', () => {
    wrapper = mount(CalendarPanel, {
      propsData: {
        type: 'year',
        defaultValue: new Date().setFullYear(11),
      },
    });
    const tds = wrapper.findAll('.mx-table-year td > div');
    tds.at(0).trigger('click');
    const expectedDate = new Date(10, 0).setFullYear(10);
    expect(wrapper.emitted().select[0][0].getTime()).toBe(expectedDate);
  });

  it('feat: active class', () => {
    wrapper = mount(CalendarPanel);
    const td = wrapper.find('.mx-table-date td:nth-child(6)');
    expect(td.classes()).not.toContain('active');
    wrapper.setProps({ value: new Date(2019, 9, 4) });
    expect(td.classes()).toContain('active');
  });

  it('feat: click prev/next month', () => {
    wrapper = shallowMount(CalendarPanel);

    const nextBtn = wrapper.find('.mx-btn-icon-right');
    const lastBtn = wrapper.find('.mx-btn-icon-left');
    const { vm } = wrapper;
    let count = 12;
    while (count--) {
      const oldYear = vm.calendarYear;
      const oldMonth = vm.calendarMonth;
      nextBtn.trigger('click');
      const newYear = vm.calendarYear;
      const newMonth = vm.calendarMonth;
      if (oldMonth === 11) {
        expect(newMonth).toBe(0);
        expect(newYear).toBe(oldYear + 1);
      } else {
        expect(newMonth).toBe(oldMonth + 1);
        expect(newYear).toBe(oldYear);
      }
    }
    count = 12;
    while (count--) {
      const oldYear = vm.calendarYear;
      const oldMonth = vm.calendarMonth;
      lastBtn.trigger('click');
      const newYear = vm.calendarYear;
      const newMonth = vm.calendarMonth;
      if (oldMonth === 0) {
        expect(newMonth).toBe(11);
        expect(newYear).toBe(oldYear - 1);
      } else {
        expect(newMonth).toBe(oldMonth - 1);
        expect(newYear).toBe(oldYear);
      }
    }
  });

  it('feat: click prev/next year', () => {
    wrapper = shallowMount(CalendarPanel, {
      propsData: {
        value: new Date(2018, 4, 5),
      },
    });
    const nextBtn = wrapper.find('.mx-btn-icon-double-right');
    const lastBtn = wrapper.find('.mx-btn-icon-double-left');
    const yearBtn = wrapper.find('.mx-btn-current-year');
    const { vm } = wrapper;
    const oldYear = vm.calendarYear;
    expect(oldYear).toBe(2018);
    nextBtn.trigger('click');
    let newYear = vm.calendarYear;
    expect(newYear).toBe(2019);
    lastBtn.trigger('click');
    newYear = vm.calendarYear;
    expect(newYear).toBe(oldYear);
    // 年视图测试
    yearBtn.trigger('click');
    expect(vm.panel).toBe('year');
    expect(vm.calendarDecade).toBe(2010);
    nextBtn.trigger('click');
    expect(vm.calendarDecade).toBe(2020);
    lastBtn.trigger('click');
    lastBtn.trigger('click');
    expect(vm.calendarDecade).toBe(2000);
  });

  const renderType = type => {
    it(`prop: type=${type}`, () => {
      wrapper = shallowMount(CalendarPanel, {
        propsData: {
          type,
          value: new Date(2019, 9, 1, 10),
        },
      });
      expect(wrapper.element).toMatchSnapshot();
    });
  };
  ['date', 'month', 'year'].forEach(renderType);

  it('prop: disabledDate', () => {
    const disabledDate = date => {
      return date < new Date(2019, 9, 1) || date > new Date(2019, 9, 20);
    };
    wrapper = mount(CalendarPanel, {
      propsData: {
        value: new Date(2019, 9, 4),
        disabledDate,
      },
    });
    const tds = wrapper.findAll('.mx-table-date td');
    for (let i = 0; i < 42; i++) {
      const td = tds.at(i);
      const classes = td.classes();
      if (i < 2 || i > 21) {
        expect(classes).toContain('disabled');
      } else {
        expect(classes).not.toContain('disabled');
      }
    }
    tds.at(1).trigger('click');
    expect(wrapper.emitted().select).toBeUndefined();
  });

  it('prop: partialUpdate', () => {
    wrapper = mount(CalendarPanel, {
      propsData: {
        value: new Date(2019, 9, 4),
        partialUpdate: true,
      },
    });
    wrapper
      .findAll('.mx-table-year td > div')
      .at(0)
      .trigger('click');
    expect(wrapper.emitted().select[0][0]).toEqual(new Date(2010, 9, 4));
    wrapper.setProps({ value: new Date(2010, 9, 4) });
    wrapper
      .findAll('.mx-table-month td > div')
      .at(0)
      .trigger('click');
    expect(wrapper.emitted().select[1][0]).toEqual(new Date(2010, 0, 4));
  });

  it('prop: defaultPanel', () => {
    wrapper = mount(CalendarPanel, {
      propsData: {
        open: true,
        type: 'month',
        defaultPanel: 'year',
      },
    });
    expect(wrapper.vm.panel).toBe('year');
  });
});
