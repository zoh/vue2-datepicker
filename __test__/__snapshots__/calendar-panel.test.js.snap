// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CalendarPanel prop: type=date 1`] = `
<div
  class="mx-calendar mx-calendar-panel-date"
>
  <div
    class="mx-calendar-header"
  >
    <button
      class="mx-btn mx-btn-text mx-btn-icon-double-left"
      type="button"
    >
      <i
        class="mx-icon-double-left"
      />
    </button>
     
    <button
      class="mx-btn mx-btn-text mx-btn-icon-left"
      type="button"
    >
      <i
        class="mx-icon-left"
      />
    </button>
     
    <button
      class="mx-btn mx-btn-text mx-btn-icon-double-right"
      type="button"
    >
      <i
        class="mx-icon-double-right"
      />
    </button>
     
    <button
      class="mx-btn mx-btn-text mx-btn-icon-right"
      type="button"
    >
      <i
        class="mx-icon-right"
      />
    </button>
     
    <span
      class="mx-calendar-header-label"
    >
      <button
        class="mx-btn mx-btn-text mx-btn-current-month"
        type="button"
      >
        
          Oct
        
      </button>
      <button
        class="mx-btn mx-btn-text mx-btn-current-year"
        type="button"
      >
        
          2019
        
      </button>
    </span>
  </div>
   
  <div
    class="mx-calendar-content"
  >
    <table-year-stub
      decade="2010"
      getcellclasses="function () { [native code] }"
      style="display: none;"
    />
     
    <table-month-stub
      getcellclasses="function () { [native code] }"
      style="display: none;"
    />
     
    <table-date-stub
      calendarmonth="9"
      calendaryear="2019"
      getcellclasses="function () { [native code] }"
      getrowclasses="function () { [native code] }"
      titleformat="YYYY-MM-DD"
    />
  </div>
</div>
`;

exports[`CalendarPanel prop: type=month 1`] = `
<div
  class="mx-calendar mx-calendar-panel-month"
>
  <div
    class="mx-calendar-header"
  >
    <button
      class="mx-btn mx-btn-text mx-btn-icon-double-left"
      type="button"
    >
      <i
        class="mx-icon-double-left"
      />
    </button>
     
    <button
      class="mx-btn mx-btn-text mx-btn-icon-left"
      style="display: none;"
      type="button"
    >
      <i
        class="mx-icon-left"
      />
    </button>
     
    <button
      class="mx-btn mx-btn-text mx-btn-icon-double-right"
      type="button"
    >
      <i
        class="mx-icon-double-right"
      />
    </button>
     
    <button
      class="mx-btn mx-btn-text mx-btn-icon-right"
      style="display: none;"
      type="button"
    >
      <i
        class="mx-icon-right"
      />
    </button>
     
    <span
      class="mx-calendar-header-label"
    >
      <button
        class="mx-btn mx-btn-text"
        type="button"
      >
        
        2019
      
      </button>
    </span>
  </div>
   
  <div
    class="mx-calendar-content"
  >
    <table-year-stub
      decade="2010"
      getcellclasses="function () { [native code] }"
      style="display: none;"
    />
     
    <table-month-stub
      getcellclasses="function () { [native code] }"
    />
     
    <!---->
  </div>
</div>
`;

exports[`CalendarPanel prop: type=year 1`] = `
<div
  class="mx-calendar mx-calendar-panel-year"
>
  <div
    class="mx-calendar-header"
  >
    <button
      class="mx-btn mx-btn-text mx-btn-icon-double-left"
      type="button"
    >
      <i
        class="mx-icon-double-left"
      />
    </button>
     
    <button
      class="mx-btn mx-btn-text mx-btn-icon-left"
      style="display: none;"
      type="button"
    >
      <i
        class="mx-icon-left"
      />
    </button>
     
    <button
      class="mx-btn mx-btn-text mx-btn-icon-double-right"
      type="button"
    >
      <i
        class="mx-icon-double-right"
      />
    </button>
     
    <button
      class="mx-btn mx-btn-text mx-btn-icon-right"
      style="display: none;"
      type="button"
    >
      <i
        class="mx-icon-right"
      />
    </button>
     
    <span
      class="mx-calendar-header-label"
    >
      <span>
        2010
      </span>
       
      <span
        class="mx-calendar-decade-separator"
      />
       
      <span>
        2019
      </span>
    </span>
  </div>
   
  <div
    class="mx-calendar-content"
  >
    <table-year-stub
      decade="2010"
      getcellclasses="function () { [native code] }"
    />
     
    <!---->
     
    <!---->
  </div>
</div>
`;
